File: python3/01.py
#! /usr/bin/env python3

def read_input():
    with open('../inputs/input01.txt') as fp:
        return [int(line) for line in fp.readlines()]

"""
Count the number of increasing pairs in the input list
"""
def part1():
    input = read_input()
    count = 0
    for i in range(len(input) - 1):
        if input[i] < input[i+1]:
            count += 1
    return count

"""
Count the number of times the sum of 3 consecutive elements increases in the input list
"""
def part2():
    input = read_input()
    count = 0
    for i in range(len(input) - 3):
        if input[i] < input[i+3]:
            count += 1
    return count

print(part1())
print(part2())

File: elixir/01.exs
defmodule Day01 do
  defp read_input do
    File.read!(Path.expand("../inputs/input01.txt"))
    |> String.split("\n")
    |> Enum.map(fn line -> line |> Integer.parse() end)
    |> Enum.map(fn {num, _} -> num end)
  end

  @doc """
  Count the number of increasing pairs in the input list
  """
  def part1 do
    input = read_input()

    input
    |> Enum.drop(1)
    |> Enum.with_index()
    |> Enum.filter(fn {_, i} ->
      Enum.at(input, i) < Enum.at(input, i + 1)
    end)
    |> length
    |> IO.inspect()
  end

  @doc """
  Count the number of times the sum of 3 consecutive elements increases in the input list
  """
  def part2 do
    input = read_input()

    input
    |> Enum.drop(3)
    |> Enum.with_index()
    |> Enum.filter(fn {_, i} ->
      # items i+1 & i+2 would be equal on each side of comparison
      Enum.at(input, i) < Enum.at(input, i + 3)
    end)
    |> length
    |> IO.inspect()
  end
end

# P1: 1559
# P1: 1600
